src/sentry/sentry_metrics/indexer/cloudspanner/cloudspanner.py

```diff
@@ -1,25 +1,65 @@
-from typing import Any, Mapping, Optional, Set
+import logging
+import re
+from datetime import datetime
+from http import HTTPStatus
+from typing import Any, Mapping, Optional, Sequence, Set
 
+import google.api_core.exceptions
 from django.conf import settings
 from google.cloud import spanner
 
 from sentry.sentry_metrics.configuration import UseCaseKey
-from sentry.sentry_metrics.indexer.base import KeyResult, KeyResults, StringIndexer
+from sentry.sentry_metrics.indexer.base import (
+    FetchType,
+    KeyCollection,
+    KeyResult,
+    KeyResults,
+    StringIndexer,
+)
 from sentry.sentry_metrics.indexer.cache import CachingIndexer, StringIndexerCache
-from sentry.sentry_metrics.indexer.id_generator import reverse_bits
+from sentry.sentry_metrics.indexer.cloudspanner.cloudspanner_model import (
+    SpannerIndexerModel,
+    get_column_names,
+)
+from sentry.sentry_metrics.indexer.id_generator import get_id, reverse_bits
+from sentry.sentry_metrics.indexer.ratelimiters import writes_limiter
 from sentry.sentry_metrics.indexer.strings import StaticStringIndexer
+from sentry.utils import metrics
 from sentry.utils.codecs import Codec
 
+logger = logging.getLogger(__name__)
+
 EncodedId = int
 DecodedId = int
 
+_INDEXER_DB_METRIC = "sentry_metrics.indexer.cloudspanner"
+_INDEXER_DB_INSERT_METRIC = "sentry_metrics.indexer.cloudspanner.insert"
+_INDEXER_DB_INSERT_ROWS_METRIC = "sentry_metrics.indexer.cloudspanner.insert_rows"
+_DEFAULT_RETENTION_DAYS = 90
 _PARTITION_KEY = "cs"
 
+# Captures the details of spanner unique constraint violation.
+# The capture group is the combination of organization_id,string,id.
+# Example error message is
+# Unique index violation on index unique_organization_string_index at index key
+# [100000,dBdOCvybLK,6866628476861885949]. It conflicts with row
+# [6866628476861885949] in table perfstringindexer_v2.
+_UNIQUE_INDEX_VIOLATION_STRING = "Unique index violation"
+_UNIQUE_VIOLATION_EXISITING_RE = re.compile(r"\[([^]]+)]")
+
```

Comment:
I don't think this is a safe way to proceed. The exception does not provide us any structured information about the error. Which means there is no contract that requires this error message to be still formatted in the same way.

I think a safer option would be to attempt write, if it fails do a read/write transactions that is guaranteed to succeed by reading the entries that you are trying to resolve + all the entries where the id conflicts:
```
WHERE string IN (x,c,v,b.....) OR id IN (1,2,3,4,5 <- the generated ids)
```

Then try to write what was not already there.
Spanner will retry the transaction if it fails, and retrying the transaction should re-execute the whole transaction logic.

Reply:
I am doing what you suggested by performing 2 transaction reads with the conditions `WHERE string IN (x,c,v,b.....)` and another with `WHERE id IN (1, 2, 3)`. Then creating a new set of rows to be inserted which excludes the results obtained from the above queries.